### PR TITLE
Renaming ReferencePolicy to ReferenceGrant

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
Renames ReferencePolicy to ReferenceGrant. Renaming this resource was originally suggested in the [community meeting on Jan 24](https://docs.google.com/document/d/1eg-YjOHaQ7UD28htdNxBR3zufebozXKyI28cl2E11tU/edit#bookmark=id.2mgmr9ue1p10), and then revisited in several additional meetings. I ran a [poll in March](https://kubernetes.slack.com/archives/CR0H13KGA/p1646244022335919) at which point "ReferenceGrant" was the most popular name. I think this has largely been accepted as something that will happen, but no one has had the time to actually make the change. With v0.5.0 around the corner, it seems like we should try to fit this change in if possible.

Getting this change in would also help with a [KEP that is relying on this resource](https://github.com/kubernetes/enhancements/pull/3295/files#r886082664) and is concerned that a rename at a later point could result in confusion/instability. 

**Does this PR introduce a user-facing change?**:
```release-note
ReferencePolicy has been renamed to ReferenceGrant
```

### CODE REVIEW

**Comment by robscott** [comment]
Adding a temporary hold so everyone has a chance to add feedback, will remove this sometime tomorrow.

/cc @shaneutt @youngnick @howardjohn @mikemorris 
/hold

**Comment by youngnick** [comment]
Looks like this was a straightforward search-and-replace, I agree with the ReferenceGrant name, and this LGTM.

/lgtm

**Review by shaneutt — APPROVED** [review]
/lgtm

File: site-src/geps/gep-709.md [hunk 1]
```diff
@@ -32,17 +32,17 @@ emerge.
 
 To ensure that Gateway API is able to safely provide this functionality, we need
 to enforce a handshake mechanism that requires resources in both namespaces to
-agree to this reference. To accomplish that, a new ReferencePolicy resource
+agree to this reference. To accomplish that, a new ReferenceGrant resource
 should be introduced.
 
-![Reference Policy](images/709-referencepolicy.png)
+![Reference Policy](images/709-referencegrant.png)
```
Comment (mikemorris) [inline site-src/geps/gep-709.md:38]: I think this image source needs to be renamed too.

File: site-src/v1alpha2/api-types/referencegrant.md [hunk 2]
```diff
@@ -1,35 +1,39 @@
-# ReferencePolicy
+# ReferenceGrant
 
-A ReferencePolicy can be used to enable cross namespace references within
+!!! note
+    This resource was originally named "ReferencePolicy". It was renamed
+    to "ReferenceGrant" to avoid any confusion with policy attachment.
```
Comment (mikemorris) [inline site-src/v1alpha2/api-types/referencegrant.md:5]: Should a note like this be added to GEP-709 too?

**Review by mikemorris — APPROVED** [review]
/lgtm

File: site-src/geps/gep-709.md (hunk 1 above)
Comment (robscott) [reply site-src/geps/gep-709.md:42]: Updated, thanks!

File: site-src/v1alpha2/api-types/referencegrant.md (hunk 2 above)
Comment (robscott) [reply site-src/v1alpha2/api-types/referencegrant.md:5]: Updated, thanks!

**Comment by robscott** [comment]
Thanks for the reviews @mikemorris, @shaneutt, and @youngnick! Updated to address feedback, PTAL.

**Comment by robscott** [comment]
Think we've got enough consensus around this, will let next LGTM merge this in.

/hold cancel

**Pull request merged**